### ISSUE TITLE
Make shard_map in_specs/out_specs definitions more flexible

### ISSUE THREAD

**Issue by jstuecker** [jax-ml/jax#34752]
It would be great to allow partition spec definitions inside of a shard_map that are agnostic to the naming of the mesh axes.

Imagine I am writing a code library that offers some shard-mapped functions that should allow a user to work with their own chosen mesh. "shard_map" already allows to pass a mesh through the context, rather than at definition time which is great. However, PartitionSpec definitions tend to be rather tied to a mesh naming convention, which makes them rather inflexible. For more flexibility, it would be great to support numerical partition specs -- e.g. P(0) referring to the first axis of a mesh. Further, it would be nice to have the option to indicate that an input/output should be mapped along all axes of a mesh -- e.g. via P(-1). 

Example
```python
import jax
import jax.numpy as jnp
from jax.sharding import PartitionSpec as P

# (1) Works, but couples the library to mesh axis names ("i","j")
@jax.shard_map(in_specs=(P(("i","j")), P(("i","j")), None),
               out_specs=P(("i","j")),
               mesh=None)
# (2) Proposed: mesh-name-agnostic via positional mesh axes
# @jax.shard_map(in_specs=(P((0,1)), P((0,1)), None), out_specs=P((0,1)), mesh=None)
# (3) Proposed: fully mesh-agnostic, shard across all mesh axes
# @jax.shard_map(in_specs=(P(-1), P(-1), None), out_specs=P(-1), mesh=None)

# User chooses a mesh in their code
mesh = jax.make_mesh((2,2), ("i","j"))
# mesh = jax.make_mesh((2,2), ("a","b")) # Could work with (2) and (3)
# mesh = jax.make_mesh((4,), ("a",)) # Could work with (3)
with jax.set_mesh(mesh):
    x = jnp.arange(4)
    y = jnp.arange(4)
    print(mul(x, y, 0.3))
```

**Comment by Ayushi6002** [comment]
This is a really interesting usability problem — especially for writing shard_map-based library code that shouldn’t depend on a particular mesh naming convention.

I agree that tying PartitionSpec directly to mesh axis names makes it harder to write reusable components. The idea of positional or mesh-agnostic specs (e.g. P((0, 1)) or a sentinel like P(-1) for “all mesh axes”) seems appealing from an API ergonomics perspective.

A few questions/thoughts as I try to understand the design space:
- Would positional specs be interpreted relative to the active mesh ordering (i.e. mesh.axis_names)?
- How would this interact with nested or partially-used meshes?
- For something like P(-1), would the intent be “replicate across all mesh axes” or “shard across all axes”, and should that be explicit?
- Are there concerns around ambiguity or silent mis-sharding when switching meshes of different ranks?

If this direction seems reasonable, I’d be interested in helping explore:
- a concrete semantic proposal
- examples showing how this improves library reusability
- or a minimal prototype / RFC-style discussion before any implementation

Happy to follow guidance on what would be most useful.

**Comment by jstuecker** [comment]
Thanks for the quick reply!
> Would positional specs be interpreted relative to the active mesh ordering (i.e. mesh.axis_names)?

This sounds sensible to me
> How would this interact with nested or partially-used meshes?

I think the rule should be that "n" gets replaced by the nth axis of the mesh and -1 gets replaced  by tuple that contains all unmentioned axes -- or by "None" if there are no unmentioned axes. (I'll put some examples below to clarify this).
> For something like P(-1), would the intent be “replicate across all mesh axes” or “shard across all axes”, and should that be explicit?

If I understand correctly, the entries of a partition spec always refer to the axes that should be sharded and replicated across all axes. A replication among all axes can already be achieved with P() or P(None). So the intend would be that P(-1) shards across all axes.
> Are there concerns around ambiguity or silent mis-sharding when switching meshes of different ranks?

Good question... I think if a shard_map has in its specs something like P(0,1) then an error should be raised if it is called with a mesh that has less than two dimensions. On the other hand, I think P(0, -1) should work with any mesh dimension >= 1 and P(-1) should even work with a 0 dimensional mesh (I think this behavior would make sense in the example above)

Here a few examples of a behavior that would seem consistent with the current API: The rule is that any integer "n" gets replaced by the nth axis of the active mesh and -1 gets replaced by a tuple that contains all axes of the mesh that were not mentioned in the same partition spec. (If that corresponds to an empty tuple (), it may be necessary to insert a None at that location or discard the value if nested)

```
* Mesh((4,), ("i",)), spec: P(-1), input (8,8)
-> Should do the same as P(("i",)) -> shards first axis among 4 devices -> mapped_input (2,8)
* Mesh((2,2), ("i","j")), spec: P(-1), input (8,8)
-> Should do the same as P(("i","j")) -> shards first axis among 2x2 devices -> mapped_input (2,8)
* Mesh((2,2), ("i","j")), spec: P(None, -1), input (8,8)
-> Should do the same as P(None, ("i","j")) -> shards second axis among 2x2 devices -> mapped_input (8,2)
* Mesh((2,2), ("i","j")), spec: P(0, -1), input (8,8)
-> Should do the same as P("i", ("j",)) -> shards axis 0 and 1 among two devices each -> mapped_input (4,4)
* Mesh((2,2,2), ("i","j","k")), spec: P(0, -1), input (8,8)
-> Should do the same as P("i",("j","k")) -> shards two axes among 2,2x2 devices -> mapped_input (4,2)
* Mesh((2,2,2,2), ("i","j","k","l")), spec: P(-1,(1,2)), input (8,8,8)
-> Should do the same as P(("i","l"),("j","k")) -> shards two axes among 2x2,2x2 devices -> mapped_input (2,2,8)
Some examples where -1 will be removed or replaced by None:
* Mesh((4,), ("i",)), spec: P((0,-1)), input (8,)
-> Should do the same as P("i") -> shards first axis among 4 devices -> mapped_input (2,)
* Mesh((4,), ("i",)), spec: P(0,-1), input (8,8)
-> Should do the same as P("i", None) -> shards first axis among 4 devices -> mapped_input (2,8)
* Mesh((4,), ("i",)), spec: P(-1,0), input (8,8)
-> Should do the same as P(None, "i") -> shards second axis among 4 devices -> mapped_input (8,2)
Error cases:
* Mesh((4,), ("i",)), spec: P(0,1), input (8,8)
-> Should raise an error, since mesh doesn't have sufficient dimensions
* Mesh((2,2), ("i","j)), spec: P(0,-1), input (8,)
-> Should raise an error, since input doesn't have two dimensions
* Mesh((2,2), ("i","j)), spec: P(-1,-1), input (8,8)
-> Should raise an error, since "-1" may appear at most once
```

**Comment by Ayushi6002** [comment]
Thanks for the detailed clarification — this makes the intended semantics much clearer.

Let me try to restate the proposal to check my understanding:

- Integer entries `n` in a PartitionSpec would be interpreted as referring to the `n`-th axis of the *active mesh* (i.e. `mesh.axis_names[n]`).
- A single `-1` may appear in a PartitionSpec and would expand to the tuple of all mesh axes *not otherwise mentioned* in that same spec.
  - If this expansion is empty, `-1` would effectively be removed or replaced by `None`, depending on nesting.
- `P(-1)` therefore means “shard across all mesh axes”, while replication across all axes remains expressible via `P()` / `P(None)`.
- Error cases arise when:
  - positional indices exceed mesh rank
  - input rank is insufficient for the expanded spec
  - `-1` appears more than once in a single spec

The examples you listed seem consistent with the existing PartitionSpec model, just with a preprocessing step that rewrites positional specs into name-based specs relative to the active mesh.

If this summary matches your intent, a possible next step could be:
- writing this as a short semantic spec / doc section (including the error cases), or
- adding tests that assert this expansion behavior before any implementation work.

I’d be happy to help with either documentation/tests or exploring a minimal prototype once there’s agreement on the semantics.

**Comment by jstuecker** [comment]
Yes, thanks, this summarizes it very well. I agree, that it could be implemented as a simple pre-processing step.
